libs/clients/smartsolutions/src/lib/smartSolutions.api.ts

```diff
@@ -0,0 +1,479 @@
+        category: LOG_CATEGORY,
+      })
+      return null
+    }
+
+    if (!res.ok) {
+      const responseErrors: PkPassServiceErrorResponse = {}
+      try {
+        const json = await res.json()
+        responseErrors.message = json?.message ?? undefined
+        responseErrors.status = json?.status ?? undefined
+        responseErrors.data = json?.data ?? undefined
+      } catch {
+        // noop
+      }
+
+      this.logger.warn('Expected 200 status for pkpass upsert', {
+        status: res.status,
+        statusText: res.statusText,
+        category: LOG_CATEGORY,
+        ...responseErrors,
+      })
+      return null
+    }
+
+    let json: unknown
+    try {
+      json = await res.json()
+    } catch (e) {
+      this.logger.warn('Unable to parse JSON for pkpass upsert', {
+        exception: e,
+        category: LOG_CATEGORY,
+      })
+      //return null
+    }
+
+    const response = json as UpsertPkPassResponse
+
+    if (response) {
+      return response
+    }
+
+    return null
+  }
+
+  async generatePkPass(
+    payload: CreatePkPassDataInput,
+    nationalId: string,
+    issuer: PkPassIssuer,
+  ) {
+    const existingPasses = await this.listPkPasses(
+      nationalId,
+      payload.passTemplateId ?? '',
+      issuer,
+    )
+
+    const containsActiveOrUnclaimed =
+      existingPasses?.passes &&
+      existingPasses.passes.some(
+        (p) =>
+          p.status === PkPassStatus.Active ||
+          p.status === PkPassStatus.Unclaimed,
+      )
+
+    if (containsActiveOrUnclaimed) {
+      const activePasses = existingPasses?.passes.filter(
+        (p) => p.status === PkPassStatus.Active,
+      )
+      if (activePasses?.length) {
+        return activePasses[0]
+      }
+
+      const unclaimedPasses = existingPasses?.passes.filter(
+        (p) => p.status === PkPassStatus.Unclaimed,
+      )
+
+      if (unclaimedPasses?.length) {
+        return unclaimedPasses[0]
+      }
+    }
+
+    const response = await this.upsertPkPass(payload, issuer)
+
+    if (response?.data?.upsertPass) {
+      return response?.data?.upsertPass
+    }
+
+    this.logger.warn(
+      'the pkpass service response did not include a generated pass',
+      {
+        serviceStatus: response?.status,
+        serviceMessage: response?.message,
+        category: LOG_CATEGORY,
+      },
+    )
+
+    return null
+  }
+
+  async generatePkPassQrCode(
+    payload: CreatePkPassDataInput,
+    nationalId: string,
+    issuer: PkPassIssuer,
+  ) {
+    const pkPass = await this.generatePkPass(payload, nationalId, issuer)
+
+    this.logger.debug(JSON.stringify(pkPass))
+    return pkPass?.distributionQRCode ?? ''
+  }
+
+  async generatePkPassUrl(
+    payload: CreatePkPassDataInput,
+    nationalId: string,
+    issuer: PkPassIssuer,
+  ) {
+    const pkPass = await this.generatePkPass(payload, nationalId, issuer)
+    return pkPass?.distributionUrl ?? ''
```

Comment:
Maybe better to return null so frontend can do something if no url ?

Reply:
yeah, true